File: source/python/memory_pool.py
# Fixed-size slab allocator / object pool simulation

SLOT_SIZE = 8
POOL_SLOTS = 8
used = [False] * POOL_SLOTS
slot_data = [None] * POOL_SLOTS
alloc_count = 0

def pool_alloc():
    global alloc_count
    for i in range(POOL_SLOTS):
        if not used[i]:
            used[i] = True
            alloc_count += 1
            print(f"alloc slot {i}")
            return i
    print("pool exhausted!")
    return None

def pool_free(slot_index):
    global alloc_count
    if used[slot_index]:
        used[slot_index] = False
        alloc_count -= 1
        print(f"freed slot {slot_index}")

def active_count():
    return sum(1 for u in used if u)

s0 = pool_alloc()
s1 = pool_alloc()
s2 = pool_alloc()
s3 = pool_alloc()
s4 = pool_alloc()

slot_data[s0] = 100
slot_data[s1] = 200

print(f"active: {active_count()}")

pool_free(1)
pool_free(3)

print(f"active after free: {active_count()}")

s5 = pool_alloc()
s6 = pool_alloc()
s7 = pool_alloc()

print(f"final active: {active_count()}")
print("pool freed")

File: source/javascript/memory_pool.js
// Fixed-size slab allocator / object pool
// Pool of 8 slots, each 8 bytes (simulated with DataView)
const SLOT_SIZE = 8;
const POOL_SLOTS = 8;
const poolBuffer = new ArrayBuffer(SLOT_SIZE * POOL_SLOTS);
const poolView = new DataView(poolBuffer);
const used = new Array(POOL_SLOTS).fill(false);
let allocCount = 0;

function poolAlloc() {
    for (let i = 0; i < POOL_SLOTS; i++) {
        if (!used[i]) {
            used[i] = true;
            allocCount++;
            console.log(`alloc slot ${i}`);
            return i; // return slot index as "pointer"
        }
    }
    console.log("pool exhausted!");
    return null;
}

function poolFree(slotIndex) {
    if (used[slotIndex]) {
        used[slotIndex] = false;
        allocCount--;
        console.log(`freed slot ${slotIndex}`);
    }
}

function activeCount() { return used.filter(Boolean).length; }

function writeI64(slot, val) { poolView.setBigInt64(slot * SLOT_SIZE, BigInt(val), true); }

const s0 = poolAlloc(), s1 = poolAlloc(), s2 = poolAlloc(), s3 = poolAlloc(), s4 = poolAlloc();
writeI64(s0, 100); writeI64(s1, 200);

console.log(`active: ${activeCount()}`);
poolFree(1); poolFree(3);
console.log(`active after free: ${activeCount()}`);

const s5 = poolAlloc(), s6 = poolAlloc(), s7 = poolAlloc();
console.log(`final active: ${activeCount()}`);
console.log("pool freed");
